url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/util/SOCGameList.java

```diff
@@ -0,0 +1,551 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ *
+ * The author of this program can be reached at thomas@infolab.northwestern.edu
+ **/
+package soc.util;
+
+import soc.disableDebug.D;
+import soc.game.SOCGame;
+import soc.game.SOCGameOption;
+import soc.message.SOCGames;
+
+import java.util.Enumeration;
+import java.util.Hashtable;
+
+
+/**
+ * A class for creating and tracking the games;
+ * contains each game's name, {@link SOCGameOption game options},
+ * {@link SOCGame} object, and mutex for synchronization.
+ *<P>
+ * In 1.1.07, moved from soc.server to soc.util package for client's use.
+ * Some methods moved to new subclass {@link soc.server.SOCGameListAtServer}.
+ * That subclass also tracks each game's clients ({@link soc.server.genericServer.StringConnection}s).
+ *<P>
+ * The client-side addGame methods allow game names to have a prefix which marks them
+ * as unjoinable by the client ({@link SOCGames#MARKER_THIS_GAME_UNJOINABLE}).
+ * If the game name has this prefix, its {@link GameInfo#canJoin} flag is set to false,
+ * as queried by {@link #isUnjoinableGame(String)}.  The prefix is stripped within addGame,
+ * and not stored as part of the game name in this list.
+ * Besides addGame, never supply this prefix to a SOCGameList method taking a game name;
+ * supply the game name without the prefix.
+ *
+ * @author Robert S. Thomas
+ */
+public class SOCGameList
+{
+    /** key = String, value = {@link GameInfo}; includes mutexes to synchronize game state access,
+     *  game options, and other per-game info
+     */
+    protected Hashtable gameInfo;
+
+    /** map of game names to {@link SOCGame} objects */
+    protected Hashtable gameData;
+
+    /** used with gamelist's monitor */
+    protected boolean inUse;
+
+    /**
+     * constructor
+     */
+    public SOCGameList()
+    {
+        gameInfo = new Hashtable();
+        gameData = new Hashtable();
+        inUse = false;
+    }
+
+    /**
+     * take the monitor for this game list; if we must wait, sleep up to 1000 ms between attempts.
+     */
+    public synchronized void takeMonitor()
+    {
+        // D.ebugPrintln("SOCGameList : TAKE MONITOR");
+
+        while (inUse)
+        {
+            try
+            {
+                wait(1000);
+            }
+            catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)